From: Jan Christoph Nordholz <hesso@pool.math.tu-berlin.de>
Date: Wed, 5 Aug 2026 16:06:32 +0200
Subject: _session_creation_core

Bug-Debian: https://bugs.debian.org/206572
Bug-Debian: https://bugs.debian.org/507817
Forwarded: no
Last-Update: 2026-05-20

Description: Use session creation timestamps for -ls ordering and -RR selection
 Affects screen's behaviour in the following situations:
 * 'screen -ls' lists available sessions sorted chronologically
 * 'screen -RR' now picks the youngest session instead of an arbitrary one
---
 socket.c | 81 ++++++++++++++++++++++++++++++++--------------------------------
 1 file changed, 40 insertions(+), 41 deletions(-)

diff --git a/socket.c b/socket.c
index bc4118d..e619ece 100644
--- a/socket.c
+++ b/socket.c
@@ -45,6 +45,7 @@
 #include <string.h>
 #include <unistd.h>
 #include <signal.h>
+#include <time.h>
 
 #if ENABLE_PAM
   #include <security/pam_appl.h>
@@ -128,15 +129,16 @@ int FindSocket(int *fdp, int *nfoundp, int *notherp, char *match)
 	int sdirlen;
 	int matchlen = 0;
 	char *name, *n;
-	int firsts = -1, sockfd;
-	char *firstn = NULL;
+	int sockfd;
 	int nfound = 0, ngood = 0, ndead = 0, nwipe = 0, npriv = 0;
 	int nperfect = 0;
 	struct sent {
 		struct sent *next;
 		int mode;
+		int fd;
 		char *name;
-	} *slist, **slisttail, *sent, *nsent;
+		time_t time_created;
+	} *slist, **slisttail, *sent, *nsent, *schosen = NULL;
 
 	if (match) {
 		matchlen = strlen(match);
@@ -217,13 +219,18 @@ int FindSocket(int *fdp, int *nfoundp, int *notherp, char *match)
 		}
 		if ((sent = malloc(sizeof(struct sent))) == NULL)
 			continue;
-		sent->next = NULL;
 		sent->name = SaveStr(name);
 		sent->mode = mode;
+		sent->fd = -1;
+		sent->time_created = SessionCreationTime(name);
+		for (slisttail = &slist; *slisttail; slisttail = &((*slisttail)->next)) {
+			if ((*slisttail)->time_created < sent->time_created)
+				break;
+		}
+		sent->next = *slisttail;
 		*slisttail = sent;
-		slisttail = &sent->next;
 		nfound++;
-		sockfd = MakeClientSocket(0);
+		sent->fd = sockfd = MakeClientSocket(0);
 		/* MakeClientSocket sets ids back to eff */
 		xseteuid(real_uid);
 		xsetegid(real_gid);
@@ -275,13 +282,15 @@ int FindSocket(int *fdp, int *nfoundp, int *notherp, char *match)
 		ngood++;
 		if (cmatch)
 			nperfect++;
-		if (fdp && (firsts == -1 || (cmatch && nperfect == 1))) {
-			if (firsts != -1)
-				close(firsts);
-			firsts = sockfd;
-			firstn = sent->name;
+		if (fdp && (!schosen ||
+			    (schosen && schosen->time_created < sent->time_created && nperfect == 0) ||
+			    (cmatch && nperfect == 1))) {
+			if (schosen && schosen->fd >= 0)
+				close(schosen->fd);
+			schosen = sent;
 		} else {
 			close(sockfd);
+			sent->fd = -1;
 		}
 	}
 	(void)closedir(dirp);
@@ -300,33 +309,23 @@ int FindSocket(int *fdp, int *nfoundp, int *notherp, char *match)
 			break;
 		}
 		for (sent = slist; sent; sent = sent->next) {
-			switch (sent->mode) {
-			case 0700:
-				printf("\t%s\t(Attached)\n", sent->name);
-				break;
-			case 0600:
-				printf("\t%s\t(Detached)\n", sent->name);
-				break;
-			case 0701:
-				printf("\t%s\t(Multi, attached)\n", sent->name);
-				break;
-			case 0601:
-				printf("\t%s\t(Multi, detached)\n", sent->name);
-				break;
-			case -1:
-				/* No trigraphs here! */
-				printf("\t%s\t(Dead ?%c?)\n", sent->name, '?');
-				break;
-			case -2:
-				printf("\t%s\t(Removed)\n", sent->name);
-				break;
-			case -3:
-				printf("\t%s\t(Remote or dead)\n", sent->name);
-				break;
-			case -4:
-				printf("\t%s\t(Private)\n", sent->name);
-				break;
-			}
+			char timestr[64];
+			if (sent->time_created == 0)
+				snprintf(timestr, sizeof(timestr), "??" "?");
+			else
+				strftime(timestr, sizeof(timestr), "%x %X", localtime(&sent->time_created));
+			printf("\t%s\t(%s)\t(%s)\n", sent->name, timestr,
+			       (sent->mode == 0700) ? "Attached" :
+			       (sent->mode == 0600) ? "Detached" :
+#ifdef MULTIUSER
+			       (sent->mode == 0701) ? "Multi, attached" :
+			       (sent->mode == 0601) ? "Multi, detached" :
+#endif
+			       (sent->mode == -1) ? "Dead ??" "?" :
+			       (sent->mode == -2) ? "Removed" :
+			       (sent->mode == -3) ? "Remote or dead" :
+			       (sent->mode == -4) ? "Private" :
+			       "Unknown");
 		}
 	}
 	if (ndead && !quietflag) {
@@ -336,9 +335,9 @@ int FindSocket(int *fdp, int *nfoundp, int *notherp, char *match)
 		else
 			Msg(0, m, ndead > 1 ? "s" : "", ndead > 1 ? "" : "es");
 	}
-	if (firsts != -1) {
-		sprintf(SocketPath + sdirlen, "/%s", firstn);
-		*fdp = firsts;
+	if (schosen) {
+		sprintf(SocketPath + sdirlen, "/%s", schosen->name);
+		*fdp = schosen->fd;
 	} else
 		SocketPath[sdirlen] = 0;
 	for (sent = slist; sent; sent = nsent) {
